here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/models/storage/HeapCacheCursor.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2017-2023 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.lib.core.models.storage;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * HeapCacheCursor reads all data (up to provided limit) ahead of time to memory, and that allows to go back and forth.
+ * It's not thread-safe in any way, {@link SeekableCursor} is designed to jump to any position of the cursor, so it doesn't really make sense to use it in parallel.
```

Comment:
Theoretically, yes, practically, I would say no. The reason is, that the feature being in the cursor are not immutable (at least not guaranteed, as we do not know the codec), therefore, when we share the copy without documenting this behavior, people can start to modify the features and get unexpected side effects. Apart, I currently do not see a single use case, where we really have multiple copies (at least not in Naksha). Actually, I would just make one implementation being MutableCursor, because you can use this one for both use cases, seeking and mutating, I see no benefit in have an In-Memory seekable only (but not mutable) cursor. I see only a use case, when you have a database that allows you to seek through results in read-only form. For me, the default implementation for both (asSeekableCursor and asMutableCursor) should simply return mutable cursors.